#164046212 Login Using Social Accounts

cob04 opened the pull request:
### What does this PR do?
Adds Social login feature for `Facebook`, `Twitter` and `Google`. Users can now login or register using their social media accounts.

### Description of task
Users should be able to use `Google`, `Facebook` and `Twitter` accounts to log in or register with Author's haven. The process should be simple and straight forward leading to good user experience.

### How has this been tested?
- [x] Unit tests

### How can this be manually tested?
- Clone this repo using `git clone https://github.com/andela/ah-legion-frontend.git`
- Checkout to the branch using `git checkout ft-social-login-164046212`
- Run `npm i` to install all dependencies
- Create a `.env`. 
- Add this line to the `.env` file  `REACT_APP_BASE_URL=https://ah-legion-staging.herokuapp.com/api/`  this references the backend, you can your localhost address if that is what you are using.
-  Add this line as well `REACT_APP_FRONTEND_URL=https://ah-legion-frontend-staging.herokuapp.com/` to your `.env` file
- Run the command `source .env` to activate your environment variables contained in  `.env`
- Navigate to the login modal by clicking the navbar login link and log in a user using the social buttons near the bottom of the popup modal.



### PT
[#164046212](https://www.pivotaltracker.com/story/show/164046212)

cob04 committed:
feature(social login): login using social accounts

- Facebook login
- Google login
- Twitter login

cob04 committed:
feature(social login): login using social accounts

- Facebook login
- Google login
- Twitter login

abulojoshua1 commented:
@cob04, kindly rebase so as to add coveralls to your PR and fix Hound CI comments.

RachelleMaina committed:
feature(create article):user can create an article
 - add create article components, container, actions and actiontypes
 - add updating while user types
 - add article publishing
 - add pretected routing
 - add tests

[Finishes #165299388]

abulojoshua1 commented:
@cob04, kindly fix the coverage drop. Also, this PR has been up for so long, make sure it is merged as soon as possible.

abulojoshua1 commented on src/components/FacebookLogin.jsx:
```diff
@@ -0,0 +1,49 @@
+      dropError: PropTypes.func.isRequired,
+      socialLogin: PropTypes.func.isRequired,
+      raiseError: PropTypes.func.isRequired,
+    };
```
@cob04, place this static proptypes at the bottom of `render()`

abulojoshua1 commented on src/components/GoogleLogin.jsx:
```diff
@@ -0,0 +1,46 @@
+import { socialLoginAction } from '../store/actions/authActions/socialLoginAction';
+
+export class GoogleLogin extends Component {
+    static propTypes = {
```
@cob04, place this at the bottom of `render()` to make the code uniform.

abulojoshua1 commented on src/components/TwitterLogin.jsx:
```diff
@@ -0,0 +1,55 @@
+    dropError: PropTypes.func.isRequired,
+    socialLogin: PropTypes.func.isRequired,
+    raiseError: PropTypes.func.isRequired,
+  };
```
@cob04, same thing here.

abulojoshua1 commented on src/css/SocialButton.css:
```diff
@@ -0,0 +1,3 @@
+.social-button {
+    margin: 2% 0 2% 0;
+}
```
@cob04, add EOF here.

abulojoshua1 commented on src/routes/index.jsx:
```diff
@@ -5,13 +5,15 @@ import Header from '../components/Header';
       <Header />
       <Route exact path="/" component={Home} />
       <Route path="/login" component={Login} />
+      <Route path="/social-login" component={SocialAccount} />
```
@cob04, you do not need a route for this, you could use the modal. If you need to deviate, notify the team so we discuss it.

cob04 replied:
thank you, this has been resolved

abulojoshua1 commented on src/store/reducers/socialLoginReducer.js:
```diff
@@ -0,0 +1,28 @@
+  }
+};
+
+export default socialLoginReducer;
```
@cob04, you do not need a different reducer for this, the whole application can use one login reducer. Kindly use @MandelaK's reducer or create one that is not specific to OAuth.

cob04 replied:
I have switched to using `loginReducer`.

abulojoshua1 commented on src/tests/actions/socialLoginAction.test.js:
```diff
@@ -0,0 +1,47 @@
+    }
+};
+
+describe("network calls", () => {
```
@cob04, make this more descriptive, use something like `OAuth network call` instead of `network calls`.

abulojoshua1 commented on src/tests/components/FacebookLogin.test.js:
```diff
@@ -0,0 +1,18 @@
+import FacebookLogin from '../../components/FacebookLogin';
+
+
+describe('GoogleComponent', () => {
```
@cob04, make this more descriptive.

abulojoshua1 commented on src/tests/components/FacebookLogin.test.js:
```diff
@@ -0,0 +1,18 @@
+      const button = component.find('button');
+      expect(component.length).toBe(1);
+  })
+})
```
@cob04, add EOF here.

abulojoshua1 commented on src/tests/components/GoogleLogin.test.js:
```diff
@@ -0,0 +1,18 @@
+    const button = component.find('button');
+    expect(component.length).toBe(1);
+  })
+})
```
same thing here.

abulojoshua1 commented on src/tests/components/GoogleLogin.test.js:
```diff
@@ -0,0 +1,18 @@
+import GoogleLogin from '../../components/GoogleLogin';
+
+
+describe('GoogleComponent', () => {
```
make this more descriptive.

abulojoshua1 commented on src/tests/components/SocialAccount.test.js:
```diff
@@ -0,0 +1,16 @@
+  })
+
+
+})
```
@cob04, add EOF here.

abulojoshua1 commented on src/tests/components/TwitterLogin.test.js:
```diff
@@ -0,0 +1,18 @@
+      const button = component.find('button');
+      expect(component.length).toBe(1);
+  })
+})
```
same thing here.

MandelaK commented on src/store/actions/authActions/socialLoginAction.js:
```diff
@@ -0,0 +1,75 @@
+      // dispatch method to remove current error
+      dispatch(socialLoginAction(token, 'google'));
+    }).catch(() => {
+      // dispatch method to raise error with name of the provider.
```
Is there a missing dispatch here? I think it's best to always dispatch an action to the store whenever a request fails

cob04 replied:
You're right, there is meant to be one there. Not implemented yet. Thank you.

MandelaK commented on src/store/actions/authActions/socialLoginAction.js:
```diff
@@ -0,0 +1,75 @@
+      const token = result.credential.accessToken;
+      dispatch(socialLoginAction(token, 'facebook'));
+    })
+    .catch(() => {
```
Please try and dispatch an action whenever a request fails

cob04 replied:
Thanks for pointing that's out, I will add those. I will use the reducers from the login feature as pointed out by @abulojoshua1

cob04 replied:
Thanks for pointing that's out, I will add those. I will use the reducers from the login feature as pointed out bt @abulo

MandelaK commented on src/tests/components/SocialAccount.test.js:
```diff
@@ -0,0 +1,15 @@
+
+describe('Social Account component', () => {
+
+  it("Should have corrent state", () => {
```
Fix typo here

cob04 replied:
will do.

ElMonstro commented on src/components/FacebookLogin.jsx:
```diff
@@ -0,0 +1,19 @@
+import { facebookLoginAction } from '../store/actions/authActions/socialLoginAction';
+
+
+export const FacebookLogin = ({ facebookLogin }) => (
```
Please validate your props @cob04

ElMonstro commented on src/tests/actions/socialLoginAction.test.js:
```diff
@@ -0,0 +1,113 @@
+      it('makes after successful call it dispatches socialloginaction', () => {
+        const expectedActions = []
+      })
+  })
```
Please add EOF line here

cob04 replied:
Thanks, on it.

ElMonstro commented on src/tests/components/GoogleLogin.test.js:
```diff
@@ -0,0 +1,38 @@
+      mapDispatchToProps(dispatch).googleLogin();
+      expect(dispatch).toHaveBeenCalled();
+    });
+  });
```
Add EOF line please

ElMonstro reviewed (approved):
LGTM

abulojoshua1 commented:
@cob04, kindly fix the coverage drop.

abulojoshua1 reviewed (approved):
LGTM
